ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -1043,53 +1082,33 @@ let rank_hosts_by_placement ~__context ~vm ~group =
     | _ ->
         []
   in
+  let hosts_without_affinity_host =
+    Db.Host.get_all ~__context
+    |> List.filter (fun host -> not (List.mem host affinity_host))
```

Comment:
This function is a bit misleading at a glance, as it's kind of treating `affinity_host` as a convenient optional type (by either being `[]` or a singleton `[ref]` containing the host). The convenience is that it matches well with list incorporation at the end, but the casing on optional remains: is the only function of the final `List.filter` to remove this potentially empty head element?

If so, it may be simpler style to be explicit and make the affinity an `option`:

```ocaml
let get_affinity_host ~__context vm =
  let r = Db.VM.get_affinity ~__context ~self:vm in
  if Db.is_valid_ref __context r then
    Some r
  else
    None

let rank_hosts_by_placement' ~__context ~vm ~group =
  let hosts = Db.Host.get_all ~__context in
  let affinity = get_affinity_host __context vm in
  let hosts_without_affinity =
    Option.fold
      ~none:hosts
      ~some:(fun host -> List.filter ((<>) host) hosts)
      affinity
  in
  let sorted_hosts =
    hosts_without_affinity
    |> rank_hosts_by_vm_cnt_in_group ~__context group
    |> List.(map (map fst))
  in
  match affinity with
  | Some host -> [host] :: sorted_hosts
  | _ -> sorted_hosts
```

The above isn't very clean either, admittedly. It could probably do with being factored a bit more.

I'm not familiar enough with the codebase to know how often one needs to query the validity of a Db reference, but maybe there could be some machinery there (suffixed `_opt` accessors) if it's common enough.

Please tell me if you feel this is overly pedantic, it just took me a few glances to understand the way that `affinity_host` is being used. Particularly because the usage of `List.mem` in the first filtering appears quadratic (if you don't know the structure of `affinity_host`) and the final filter appears to exist purely to implicitly case out the optional.

Reply:
Those are so nice suggestions. You are too kind. Please don't use "pedantic" in the review of code again.